contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,104 @@
+#include <vesting.token/vesting.token.hpp>
+
+namespace vestingtoken {
+
+    void vestingToken::updatedate(eosio::time_point_sec new_start_date) {
+        require_auth(get_self());
+        start_date start_date(get_self(), get_self().value);
+        start_date.set(new_start_date, get_self());
+    }
+
+    void vestingToken::assigntokens(eosio::name holder, eosio::asset amount, vesting_category category) {
+        // Only the contract owner can call this function
+        require_auth(get_self());
+
+        // Check the symbol is correct and valid
+        auto sym = amount.symbol;
+        eosio::check(sym == system_resource_currency, "Symbol does not match system resource currency");
+        eosio::check(sym.precision() == system_resource_currency.precision(), "Symbol precision does not match");
+        eosio::check(amount.amount > 0, "Amount must be greater than 0");
+
+        // Create a new vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+        auto iter = vesting_table.end();
+
+        // Check the number of rows for the user
+        uint32_t num_rows = 0;
+        for (auto itr = vesting_table.begin(); itr != vesting_table.end(); ++itr) {
+            if (itr->holder == holder) {
+                ++num_rows;
+                if (num_rows > 20) {
+                    eosio::check(false, "Cannot purchase tokens more than 20 times.");
+                }
+            }
+        }
+
+        vesting_table.emplace(get_self(), [&](auto& row) {
+            row.holder = holder;
+            row.total_allocated = amount;
+            row.tokens_claimed = asset(0, amount.symbol);
+            row.allocated = eosio::current_time_point();
+            row.vesting_category_type = category;
+        });
+    }
+    void vestingToken::withdraw(eosio::name holder) {
+        require_auth(holder);
+
+        // Get the vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+        eosio::check(start_date(get_self(), get_self().value).exists(), "Start date not set");
+
+        for (auto iter = vesting_table.begin(); iter != vesting_table.end(); ++iter) {
+            const vested_allocation& vesting_schedule = *iter;
+
+            vesting_category category = vesting_schedule.vesting_category_type;
+
+            eosio::time_point_sec start_delay_seconds = eosio::time_point_sec(category.start_delay_days * SECONDS_IN_DAY);
+            eosio::time_point_sec start_date_value = start_date(get_self(), get_self().value).get();
+            eosio::time_point_sec vesting_start = start_date_value + start_delay_seconds.sec_since_epoch() + vesting_schedule.allocated.sec_since_epoch();
```

Comment:
global
- sales start date: absolute
- launch date: absolut

vesting schedule
- uint32_t  cliff_period_days;  relative
 - uint32_t  start_delay_days;  relative
 - uint32_t  vesting_period_days;  relative

allocation:
- allocated: relative to sales start date

```c++
allocate_tokens() {_

allocated_after_sales_start_seconds = now.sec_since_epoch() - global_sales_sate.sec_since_epoch()
}

withdraw() {
  const time_point_sec now = eosio::current_time_point();
  for loop over holder {
    uint32_t vesting_start_since_epoch = global_launch_date.sec_since_epoch()
      + allocated_after_sales_start_seconds
      + category.start_delay_days*SECONDS_IN_DAY
   uint32_t cliff_start_since_epoch = vesting_start_since_epoch + category.cliff_start_day*SECONDS_IN_DAY

eosio::check(now.sec_since_epoch() >= cliff_start_since_epoch , "Vesting period after cliff has not started");
  }
}
```